src/eftest/report/junit.clj

```diff
@@ -0,0 +1,156 @@
+(ns eftest.report.junit
+  "A test reporter that outputs JUnit-compatible XML."
+  (:require [clojure.java.io :as io]
+            [clojure.stacktrace :as stack]
+            [clojure.test :as test]
+            [eftest.report :refer [*context*]]))
+
+(def ^:dynamic *junit-path*
+  "Path for JUnit XML output"
+  "test-results.xml")
+
+;; XML generation based on junit.clj
+
+(def ^{:private true}
+     escape-xml-map
+     (zipmap "'<>\"&" (map #(str \& % \;) '[apos lt gt quot amp])))
+
+(defn- escape-xml [text]
+  (apply str (map #(escape-xml-map % %) text)))
+
+(defn start-element
+  [tag & [attrs]]
+  (print (str "<" tag))
+  (if (seq attrs)
+    (doseq [[key value] attrs]
+      (print (str " " (name key) "=\"" (escape-xml value) "\""))))
+  (print ">"))
+
+(defn element-content
+  [content]
+  (print (escape-xml content)))
+
+(defn finish-element
+  [tag]
+  (print (str "</" tag ">")))
+
+(defn test-name
+  [vars]
+  (apply str (interpose "."
+                        (reverse (map #(:name (meta %)) vars)))))
+
+(defn package-class
+  [name]
+  (let [i (.lastIndexOf name ".")]
+    (if (< i 0)
+      [nil name]
+      [(.substring name 0 i) (.substring name (+ i 1))])))
+
+(defn start-case
+  [name classname]
+  (start-element 'testcase {:name name :classname classname}))
+
+(defn finish-case
+  []
+  (finish-element 'testcase))
+
+(defn suite-attrs
+  [package classname]
+  (let [attrs {:name classname}]
+    (if package
+      (assoc attrs :package package)
+      attrs)))
+
+(defn start-suite
+  [name]
+  (let [[package classname] (package-class name)]
+    (start-element 'testsuite (suite-attrs package classname))))
+
+(defn finish-suite
+  []
+  (finish-element 'testsuite))
+
+(defn message-el
+  [tag message expected-str actual-str file line]
+  (start-element tag (if message {:message message} {}))
+  (element-content
+   (let [detail (apply str (interpose
+                            "\n"
+                            [(str "expected: " expected-str)
+                             (str "  actual: " actual-str)
+                             (str "      at: " file ":" line)]))]
+     (if message (str message "\n" detail) detail)))
+  (finish-element tag)
+  (println))
+
+(defn failure-el
+  [{:keys [message expected actual file line]}]
+  (message-el 'failure message (pr-str expected) (pr-str actual) file line))
+
+(defn error-el
+  [{:keys [message expected actual file line]}]
+  (message-el 'error
+              message
+              (pr-str expected)
+              (if (instance? Throwable actual)
+                (with-out-str (stack/print-cause-trace actual test/*stack-trace-depth*))
+                (prn actual))
+              file line))
+
+;; The reporting function
+
+(defmacro with-junit-out [& body]
```

Comment:
Why not use `with-test-out` instead? We can then use the `*test-out*` var to determine where we want to output.